mlrun/artifacts/document.py

```diff
@@ -0,0 +1,313 @@
+                for char in src_path[current_pos : match.start()]:
+                    result.append(hex(ord(char))[2:].zfill(2))
+
+                # Add the valid sequence
+                result.append(match.group())
+                current_pos = match.end()
+
+            # Handle any remaining characters after the last match
+            for char in src_path[current_pos:]:
+                result.append(hex(ord(char))[2:].zfill(2))
+
+            resolved_path = "".join(result)
+
+            artifact_key = artifact_key.replace("%%", resolved_path)
+
+        return artifact_key
+
+
+class DocumentArtifact(Artifact):
+    """
+    A specific artifact class inheriting from generic artifact, used to maintain Document meta-data.
+
+    Methods:
+        to_langchain_documents(splitter): Create LC documents from the artifact.
+        collection_add(collection_id): Add a collection ID to the artifact.
+        collection_remove(collection_id): Remove a collection ID from the artifact.
+    """
+
+    class DocumentArtifactSpec(ArtifactSpec):
+        _dict_fields = ArtifactSpec._dict_fields + [
+            "document_loader",
+            "collections",
+            "original_source",
+        ]
+
+        def __init__(
+            self,
+            *args,
+            **kwargs,
+        ):
+            super().__init__(*args, **kwargs)
+            self.document_loader = None
+            self.collections = set()
+            self.original_source = None
+
+    """
+    A specific artifact class inheriting from generic artifact, used to maintain Document meta-data.
+    """
+
+    kind = "document"
+
+    METADATA_SOURCE_KEY = "source"
+    METADATA_ORIGINAL_SOURCE_KEY = "original_source"
+    METADATA_CHUNK_KEY = "mlrun_chunk"
+    METADATA_ARTIFACT_URI_KEY = "mlrun_object_uri"
+    METADATA_ARTIFACT_TARGET_PATH_KEY = "mlrun_target_path"
+
+    def __init__(
+        self,
+        key=None,
+        document_loader: DocumentLoaderSpec = DocumentLoaderSpec(),
+        **kwargs,
+    ):
+        super().__init__(key, **kwargs)
+        self.spec.document_loader = document_loader.to_str()
+        if "src_path" in kwargs:
+            self.spec.original_source = kwargs["src_path"]
+
+    @property
+    def spec(self) -> DocumentArtifactSpec:
+        return self._spec
+
+    @spec.setter
+    def spec(self, spec):
+        self._spec = self._verify_dict(
+            spec, "spec", DocumentArtifact.DocumentArtifactSpec
+        )
+        # _verify_dict doesn't handle set, so we need to convert it back
+        if isinstance(self._spec.collections, str):
+            self._spec.collections = ast.literal_eval(self._spec.collections)
+
+    @property
+    def inputs(self):
+        # To keep the interface consistent with the project.update_artifact() when we update the artifact
+        return None
+
+    @property
+    def source(self):
+        return generate_artifact_uri(self.metadata.project, self.spec.db_key)
```

Comment:
I'm not a fan of hiding logic behind a property. Would much prefer changing this to `get_source()` which hints to users that something is being done behind the scenes.

Reply:
will fix